Feature - redis sentinel support

rolandg opened the pull request:
Hi,
I'm new to go. Any advice is more than welcome.
  Fixes #38 

### What is the purpose of this pull request?
Introduce redis sentinel support.

### What changes did you make? (overview)
I used github.com/FZambia/sentinel to make this redis sentinel compatible.
Not sure why I had to add `CGO_ENABLED=0` to get the binary working inside the alpine image.

I had to add `redis_sentinel_enabled`, `redis_sentinel`, `redis_password`, `redis_master_name` as parameters. Didn't want to break previous configurations but now the user needs to provide the password different depending on if he is using redis sentinel or not. Generally I'd go for the dedicated password field instead of the password included in the url.

Currently testing this. 

TODO: tests
Let me konw what you think
Thanks

palkan commented on cli/cli.go:
```diff
@@ -59,6 +59,11 @@ func init() {
 	fs.StringVar(&defaults.RedisURL, "redis_url", redisDefault, "")
 	fs.StringVar(&defaults.RedisChannel, "redis_channel", "__anycable__", "")
 
+	fs.BoolVar(&defaults.RedisSentinelEnabled, "redis_sentinel_enabled", false, "")
```
Can we simplify the configuration by re-using `redis_url` for master name and password?
And also make `redis_sentinel_enabled` by checking whether the list of sentinels is provided or not?

Thus, only a single new parameter would be required.

P.S. That's the way Ruby client handles this: https://bugs.ruby-lang.org/versions/36

palkan commented on pubsub/redis.go:
```diff
@@ -65,10 +75,79 @@ func (s *RedisSubscriber) Start() error {
+
+		defer sntnl.Close()
+
+		pool := &redis.Pool{
```
Do we need a pool here? We're only using a single connection.

We only need a role check right after the `Dial`.

palkan commented on pubsub/redis.go:
```diff
@@ -65,10 +75,79 @@ func (s *RedisSubscriber) Start() error {
+		c = pool.Get()
+
+	} else {
+		c, err = redis.DialURL(s.url)
```
Can't we re-use this functionality by generating an URL from `sntnl.MasterAddr()`?

palkan commented on pubsub/redis.go:
```diff
@@ -65,10 +75,79 @@ func (s *RedisSubscriber) Start() error {
+		s.log.Debugf("Redis sentinel enabled")
+		s.log.Debugf("Redis sentinel parameters:  sentinels: %s,  masterName: %s", s.sentinels, s.masterName)
+		sentinels := strings.Split(s.sentinels, ",")
+		sntnl := &sentinel.Sentinel{
```
I think, we can build a sentinel once (in `Start()` function), not for every connection attempt (the `listen` is called for every reconnection).

rolandg commented:
Hi. 
Thank You for your input. 
Updated things according to your comments.
I have also added periodic sentinel discovery.

Will now move this to the correct branch. Not sure if I have to open a new pr.

palkan commented on pubsub/redis.go:
```diff
@@ -48,45 +42,25 @@ func NewRedisSubscriber(node *node.Node, url string, sentinelsEnabled bool, sent
 // Start connects to Redis and subscribes to the pubsub channel
 func (s *RedisSubscriber) Start() error {
 	// Check that URL is correct first
```
I think, we should update the comment as well)

palkan commented on pubsub/redis.go:
```diff
@@ -38,13 +42,79 @@ func NewRedisSubscriber(node *node.Node, url string, channel string) RedisSubscr
+		defer sntnl.Close()
+
+		// Periodically discover new Sentinels.
+		go func() {
```
We need to take care (=stop) of this groutine when we exiting the `Start()` method (for example, when we exhausted all the reconnect attempts).

We can use `context.WithCancel` or simply a channel + `select`.

rolandg replied:
Thanks again. Not sure if i did that right. Could you please check https://github.com/anycable/anycable-go/pull/95/commits/c9be92e3d4b442035be3d35dd68efddfc9268825
Cheers

palkan replied:
LGTM 👍

palkan commented on CHANGELOG.md:
```diff
@@ -35,6 +35,8 @@ Disabled by default due to the experimental status in [Gorilla](https://github.c
 
 Now you can specify only the part of the version, e.g. `anycable-go:1.0` instead of the full `anycable-go:v1.0.0`.
 
+- Added redis sentinel support([@rolandg][])
```
```suggestion
- Added Redis Sentinel support([@rolandg][])
```

sponomarev commented on pubsub/redis.go:
```diff
@@ -38,13 +42,79 @@ func NewRedisSubscriber(node *node.Node, url string, channel string) RedisSubscr
 		return err
 	}
 
+	var sntnl *sentinel.Sentinel
```
What is the reason for that explicit declaration instead of good ol' shorthand syntax?

rolandg replied:
that reason is gone now. thanks. updated it

sponomarev commented on pubsub/redis.go:
```diff
@@ -38,13 +42,79 @@ func NewRedisSubscriber(node *node.Node, url string, channel string) RedisSubscr
+			}
+			s.log.Debugf("Got master address from sentinel: %s", masterAddress)
+
+			if password == "" {
```
I don't get the idea of reconstruction `s.url` if the propper URL is already there. Especially, doing that we loose url schema which can contain `rediss` for TLS-secured instances.

rolandg replied:
What do you mean with: the proper url is already there? We go from `redis://[:pass@]<mastername>` to `redis://[:pass@]<actual ip of the current master>`.  Didn't know scheme can change. I'll use the pased one here instead of hadcoding it.

rolandg replied:
or should we just replace host of the redisUrl object and use string here? guess thats what you meant

rolandg replied:
did that

rolandg commented on pubsub/redis.go:
```diff
@@ -65,12 +135,22 @@ func (s *RedisSubscriber) Start() error {
 func (s *RedisSubscriber) listen() error {
-	c, err := redis.DialURL(s.url)
+
+	var c redis.Conn
```
that reason is gone. updated it thanks

rolandg commented on pubsub/redis.go:
```diff
@@ -38,13 +42,79 @@ func NewRedisSubscriber(node *node.Node, url string, channel string) RedisSubscr
+				s.log.Warn("Failed to discover sentinels")
+			}
+			for {
+				<-time.After(30 * time.Second)
```
I added it as a parameter

rolandg commented on pubsub/redis.go:
```diff
@@ -17,34 +20,109 @@ const (
+
+			if err != nil {
+				s.log.Warn("Failed to get master address from sentinel.")
+				return err
```
Thanks for looking into this. Need to check the demo app. I've been testing this in our test environment before but since I rebased to 1.0-dev it stopped working. Guess I'd have to update also the ruby side.
I'll test this with the demo app as soon I got some spare time.

palkan replied:
Yeah, you need `1.0-dev` branch of `anycable` Ruby gem

rolandg replied:
Hi. 
Unfortunately I cannot reproduce the issue with reconnection after I stop the master.
[Here](https://gist.github.com/rolandg/13eba4174f62ff0b24d1a0d190609a88) is a gist with the logs. 
I'll check the auto retry when sentinels are not yet available.

rolandg replied:
Ah I see. You are using the pause command. 
That does trigger a failover but doesn't interrupt the connection from anycable to the master. If you use docker stop the failover works.

rolandg replied:
Yes we can move it that would be more clean. Would you use function parameters for `redisUrl` and `sntnl` or write it to the `RedisSubscriber` object?

An alternative would be just drop the return. This works but does an unnecessary connection attempt to localhost:6379:
```
anycable_1        | W 2020-05-04T15:18:17.136Z context=pubsub Failed to get master address from sentinel.
anycable_1        | D 2020-05-04T15:18:17.136Z context=pubsub Got master address from sentinel: 
anycable_1        | W 2020-05-04T15:18:37.139Z context=pubsub Redis connection failed: dial tcp 127.0.0.1:6379: connect: connection refused
anycable_1        | I 2020-05-04T15:18:37.139Z context=pubsub Next Redis reconnect attempt in 1m21s
```

rolandg commented:
Used the wrong thread. Here again the info about the failover:

Hi.
Unfortunately I cannot reproduce the issue with reconnection after I stop the master.
Here is a gist with the logs. (https://gist.github.com/rolandg/13eba4174f62ff0b24d1a0d190609a88)

Ah I see. You are using the pause command.
That does trigger a failover but doesn't interrupt the connection from anycable to the master. If you use docker stop the failover works. Not sure what that means for a real world failover but I guess in that case the connection would be interrupted.

palkan commented:
> That does trigger a failover but doesn't interrupt the connection from anycable to the master

Interesting. And our [ping](https://github.com/anycable/anycable-go/blob/63a541ba08b361767915571b5331f4c9fdc86f6b/pubsub/redis.go#L194) couldn't catch this as well?

rolandg commented:
Just tried it and found that you can still ping a paused container.

palkan commented:
> Just tried it and found that you can still ping a paused container with the cli tool.

Thanks for checking. I'll take another look later this week and merge this if everything is OK.

rolandg commented:
Thanks a lot.
Do you have a rough estimate when you want to release 1.0 in general?

palkan commented:
> Do you have a rough estimate when you want to release 1.0 in general?

ASAP 🙂 That's currently my top open-source priority. I expect a release candidate be ready in the end of May, and the final release is no later than the 1st of July.
